clis/teliod/src/daemon.rs

```diff
@@ -47,32 +65,112 @@ fn telio_task(
         None,
     )?;
 
+    // Keep track of current meshnet states
+    let mut state = TelioTaskStates::default();
+    // Mock the interface configuration for tests
+    let sys_config: Box<dyn SystemConfigurator> = if auth_token.as_str().eq(EMPTY_TOKEN) {
+        // And some fake IP address
+        state.interface_ip_address = Some(IpAddr::V4(Ipv4Addr::new(100, 100, 100, 100)));
+        Box::new(MockSystemConfigurator::new(false))
+    } else {
+        Box::new(IfconfigSystemConfigurator)
+    };
+
     // TODO: This is temporary to be removed later on when we have proper integration
     // tests with core API. This is to not look for tokens in a test environment
     // right now as the values are dummy and program will not run as it expects
     // real tokens.
     if !auth_token.as_str().eq("") {
         start_telio(&mut telio, node_identity.private_key, interface_name)?;
-        task_retrieve_meshmap(node_identity, auth_token, tx_channel);
+
+        // Auto enable meshnet for the first time
+        if auto_start_meshnet {
+            task_enable_meshnet(tx_channel.clone());
```

Comment:
this sounds a bit confusing to me. 

we have:
- `task_enable_meshnet` which actually loops a `SetMeshnet` command back to the daemon. I was expecting a `telio = Device::new(..)` and `telio.start()` instead?
- `TelioTaskCmd::SetMeshnet` should enable/disable libtelio, instead is just updating the meshmap?
- `task_retrieve_meshmap` not only retrieves the meshmap but also updates it, also enabling meshnet?

Reply:
This is not a very objective comment, I just have the feeling that this is overengineered and could be simplified by using functions with a more straightforward definition. 

I'm finding that is confusing if we are starting telio or setting a valid meshnet config.

Reply:
Starting and stopping telio itself is probably much more costly in terms of operations (starting the runtime creating the tunnel interface, sockets etc) than simply setting the meshnet map. 
Correct me if I'm wrong, but I recall the [recommended](https://github.com/NordSecurity/libtelio/blob/main/src/doc/introduction.md#creating-a-meshnet) way of enabling and disabling meshnet was through `set_config()`, with disabling simply being `device.set_config(&None)`.

Also if we couple starting telio device to enabling meshnet, we would not be able to support the use case of connecting to an VPN exit node without meshnet being enabled first. But maybe we should clarify with PO if we would ever want to support this use case for teliod, at least in the desktop apps it is a standard use-case.

Reply:
Hmm `task_retrieve_meshmap` is also set as the callback to the `NotificationCenter` so it has to retrieve and update meshmap in one step. Though there is the check to prevent it from setting meshmap if meshnet is disabled.
```
if state.meshnet_is_enabled {
  telio.set_config(&Some(map)) {}
}
```

I tried to decouple it in the beginning but then it would require even more loop back commands that made the flow even more complicated to follow.. What changes would make it clearer?

Reply:
> Starting and stopping telio itself is probably much more costly in terms of operations (starting the runtime creating the tunnel interface, sockets etc) than simply setting the meshnet map.
Correct me if I'm wrong, but I recall the [recommended](https://github.com/NordSecurity/libtelio/blob/main/src/doc/introduction.md#creating-a-meshnet) way of enabling and disabling meshnet was through set_config(), with disabling simply being device.set_config(&None).

The usecase I see to stop libtelio through `teliod` is for instace if we stop the meshnet to enable a TelioFeature or change the config, in that case it will be not possible as you'll have to restart `telio::Device`. I think the cost of doing that should be somewhat trivial in this scenario.

> Also if we couple starting telio device to enabling meshnet, we would not be able to support the use case of connecting to an VPN exit node without meshnet being enabled first. But maybe we should clarify with PO if we would ever want to support this use case for teliod, at least in the desktop apps it is a standard use-case.

For QNAP usecase, VPN/exit-node feature must be disabled. Nevertheless, in that scenario I can't see why we shouldn't be able to do both `start telio > connect to exit node` and `start telio > connect to meshnet`.

Reply:
Hmm `TelioFeature` is part of `Device::new()` and not `telio.start()`, so we would have to re-create the whole device not just re-start it. We would have to keep track of more states, I'm worried if that would make things even harder to follow.

Is it part of the requirements that we dynamically change the `TelioFeatures` during run time? 

Also the config is passed as an argument to `teliod`, so we would have to restart whole of `teliod` either way if the config changes.

Reply:
Moved enabling meshnet out of this PR